src/debug/debug_stream/debug_stream_slot.c

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Intel Corporation.
+
+#include <zephyr/logging/log.h>
+#include <zephyr/spinlock.h>
+#include <adsp_debug_window.h>
+#include <adsp_memory.h>
+#include <sof/common.h>
+#include <rtos/string.h>
+#include "debug_stream.h"
+#include "debug_stream_slot.h"
+
+LOG_MODULE_REGISTER(debug_strem_slot);
+
+struct cpu_mutex {
+	struct k_mutex m;
+} __aligned(CONFIG_DCACHE_LINE_SIZE);
+
+static struct cpu_mutex cpu_mutex[CONFIG_MP_MAX_NUM_CPUS];
+
+static const int slot_num = CONFIG_SOF_DEBUG_STREAM_SLOT_NUMBER;
+
+static struct debug_stream_slot_hdr *get_my_slot(void)
+{
+	return (struct debug_stream_slot_hdr *)ADSP_DW->slots[slot_num];
+}
+
+static
+struct debug_stream_circular_buf *get_circular_buffer(struct debug_stream_section_descriptor *desc,
+						      unsigned int core)
+{
+	struct debug_stream_slot_hdr *hdr = get_my_slot();
+	void *ptr;
+
+	if (hdr->hdr.magic != DEBUG_STREAM_IDENTIFIER) {
+		LOG_ERR("Debug stream slot not initialized.");
+		return NULL;
+	}
+
+	*desc = hdr->section_desc[core];
+	LOG_INF("Section %u (desc %u %u %u)", core, desc->core_id, desc->buf_words, desc->offset);
+
+	return (struct debug_stream_circular_buf *) (((uint8_t *)hdr) + desc->offset);
+}
+
+int debug_stream_slot_send_record(struct debug_stream_record *rec)
+{
+	struct debug_stream_section_descriptor desc;
+	struct debug_stream_circular_buf *buf = get_circular_buffer(&desc, arch_proc_id());
+	uint32_t record_size = rec->size;
+	uint32_t record_start, buf_remain;
+
+	LOG_DBG("Sending record %u id %u len %u\n", rec->serial, rec->id, rec->size);
+
+	if (!buf)
+		return -ENODEV;
+
+	if (rec->size >= desc.buf_words) {
+		LOG_ERR("Record too big %u >= %u (desc %u %u %u)", rec->size,
+			desc.buf_words, desc.core_id, desc.buf_words, desc.offset);
+		return -ENOMEM;
+	}
+	k_mutex_lock(&cpu_mutex[arch_proc_id()].m, K_FOREVER);
+
+	rec->serial = buf->next_serial++;
+	rec->size = record_size + 1; /* +1 for size at the end of recored */
+	record_start = buf->w_ptr;
+	buf->w_ptr = (record_start + record_size) % desc.buf_words;
+	buf_remain = desc.buf_words - record_start;
+	if (buf_remain < record_size) {
+		uint32_t rec_remain = record_size - buf_remain;
+
+		assert(!memcpy_s(&buf->data[record_start], buf_remain * sizeof(uint32_t),
```

Comment:
Please no code in asserts. If you build with asserts disabled, the functionality will not be the same.